rust/libnewsboat/src/utils.rs

```diff
@@ -499,27 +501,50 @@ pub fn get_command_output(cmd: &str) -> String {
         .unwrap_or_else(|_| String::from(""))
 }
 
-// This function assumes that the user is not interested in command's output (not even errors on
-// stderr!), so it redirects everything to /dev/null.
+// This function assumes that the user is not interested in the command's output (not even errors
+// on stderr!), so it will close the spawned command's fds.
+// This used to be a simple std::process::Command::Spawn(), but this caused child processes to not
+// be reaped, hence the addition of double forking to this function.
+// Spawn() was replaced with a direct call to fork+execvp because it interacted badly with the
+// fork() call.
 pub fn run_command(cmd: &str, param: &str) {
-    let child = Command::new(cmd)
-        .arg(param)
-        // Prevent the command from blocking Newsboat by asking for input
-        .stdin(Stdio::null())
-        // Prevent the command from botching the screen by printing onto it.
-        .stdout(Stdio::null())
-        .stderr(Stdio::null())
-        .spawn();
-    if let Err(error) = child {
-        log!(
-            Level::Debug,
-            "utils::run_command: spawning a child for \"{}\" failed: {}",
-            cmd,
-            error
-        );
-    }
+    unsafe {
+        let forked_pid = fork();
+
+        match forked_pid {
+            -1 => {
+                // Parent process, fork failed.
+                log!(
+                    Level::Debug,
+                    "utils::run_command: failed to fork. Aborting run_command."
+                );
+                return;
+            }
+            0 => {
+                // Child process, continue and spawn the command.
+            }
+            _ => {
+                // Parent process, fork succeeded: reap child and return.
+                let mut status = 0;
+                waitpid(forked_pid, &mut status, 0);
+                return;
+            }
+        }
 
-    // We deliberately *don't* wait for the child to finish.
+        if fork() == 0 {
+            // Grand-child, close our fds to avoid clobbering the screen and exec the command
+            close(0);
+            close(1);
+            close(2);
+            let c_cmd = CString::new(cmd).unwrap();
```

Comment:
We strive to avoid `unwrap()` and other sources of panics, because they crash the entire application. In `libnewsboat-ffi` crate, we usually write a comment explaining why a certain line won't panic, but the justification often relies on the fact that the data crossed over the FFI boundary, which e.g. guarantees that there are no NUL bytes inside a given string. We can't say the same here, since we're given an `&str`.

I think the best approach would be to just handle the possibility of `CString::new()` returning an `Err`. Note that in Rust blocks can return a value (like in ML or Ruby), so you can write:

```rust
let c_cmd = match CString::new(cmd) {
    Ok(cstring) => cstring,
    Err(error) => log!(LogLevel::USER_ERROR, "…"),
};

// use c_cmd here
```